cpu/lm4f120/periph/gpio.c

```diff
@@ -0,0 +1,316 @@
+/*
+ * Copyright (C) 2015 Marc Poulhiès
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser General
+ * Public License v2.1. See the file LICENSE in the top level directory for more
+ * details.
+ */
+
+/**
+ * @ingroup     cpu_lm4f120
+ * @{
+ *
+ * @file
+ * @brief       Low-level GPIO driver implementation
+ *
+ * @author      Marc Poulhiès <dkm@kataplop.net>
+ *
+ * @}
+ */
+
+
+#include <stdio.h>
+
+#include "cpu.h"
+#include "cpu_conf.h"
+#include "periph/gpio.h"
+#include "periph_conf.h"
+
+#define ENABLE_DEBUG (0)
+#include "debug.h"
+
+/**
+ * @brief     Extract the pin number of the given pin
+ */
+static inline uint8_t _pin_num(gpio_t pin)
+{
+    return (pin & 0x0f);
+}
+
+/**
+ * @brief     Extract the port number of the given pin
+ */
+static inline uint8_t _port_num(gpio_t pin)
+{
+    return (pin >> 4);
+}
+
+static const uint32_t _sysctl_port_base[] = {
+    SYSCTL_PERIPH_GPIOA,
+    SYSCTL_PERIPH_GPIOB,
+    SYSCTL_PERIPH_GPIOC,
+    SYSCTL_PERIPH_GPIOD,
+    SYSCTL_PERIPH_GPIOE,
+    SYSCTL_PERIPH_GPIOF,
+};
+
+static const uint32_t _port_base[] = {
+    GPIO_PORTA_BASE,
+    GPIO_PORTB_BASE,
+    GPIO_PORTC_BASE,
+    GPIO_PORTD_BASE,
+    GPIO_PORTE_BASE,
+    GPIO_PORTF_BASE,
+};
+
+static const uint32_t _int_assign[] = {
+    INT_GPIOA,
+    INT_GPIOB,
+    INT_GPIOC,
+    INT_GPIOD,
+    INT_GPIOE,
+    INT_GPIOF,
+};
+
+#define NUM_OF_PORT 6
+#define NUM_OF_PINS 8
+
+typedef struct {
+    gpio_cb_t cb;       /**< callback called from GPIO interrupt */
+    void *arg;          /**< argument passed to the callback */
+} gpio_state_t;
+
+static gpio_state_t gpio_config[NUM_OF_PORT][NUM_OF_PINS];
+
+static inline uint16_t _port_addr(gpio_t pin)
+{
+    const uint8_t port_num = _port_num(pin);
+    const uint32_t port_addr = _port_base[port_num];
+
+    return port_addr;
+}
+
+int gpio_init(gpio_t pin, gpio_dir_t dir, gpio_pp_t pullup)
+{
+    const uint8_t port_num = _port_num(pin);
+    const uint32_t port_addr = _port_base[port_num];
+    const uint8_t pin_num = _pin_num(pin);
+    const uint32_t sysctl_port_base = _sysctl_port_base[port_num];
+    const unsigned long pin_bit = 1 << pin_num;
+    unsigned long pad_config;
+
+    DEBUG("Init GPIO: port %c, %d\n", 'A' + port_num, pin_num);
+    DEBUG("Sysctl %" PRIx32 "\n", sysctl_port_base);
+
+    ROM_SysCtlPeripheralEnable(sysctl_port_base);
+
+    HWREG(port_addr+GPIO_LOCK_R_OFF) = GPIO_LOCK_KEY;
+    HWREG(port_addr+GPIO_CR_R_OFF) |=  pin_bit;
+    HWREG(port_addr+GPIO_DEN_R_OFF) |= pin_bit;
+    HWREG(port_addr+GPIO_LOCK_R_OFF) = 0;
+
+
+    switch(pullup){
+    case GPIO_PULLDOWN:
+        pad_config = GPIO_PIN_TYPE_STD_WPD;
+        break;
+    case GPIO_PULLUP:
+        pad_config = GPIO_PIN_TYPE_STD_WPU;
+        break;
+    case GPIO_NOPULL:
+    default:
+        pad_config = GPIO_PIN_TYPE_STD;
+        break;
+    }
```

Comment:
Better: Just re-define the values for the `gpio_pp_t` in the `periph_cpu.h` like this:
```c
#define HAVE_GPIO_PP_T
typedef enum {
    GPIO_NOPULL = GPIO_PIN_TYPE_STD,        
    GPIO_PULLUP = GPIO_PIN_TYPE_STD_WPU,        
    GPIO_PULLDOWN = GPIO_PIN_TYPE_STD_WPD
} gpio_pp_t;
```
and you can delete this complete switch/case statement and use the `pullup` variable directly in the next line...